New feature: marking rooms as unread [MSC2867]

alangecker opened the pull request:
implements https://github.com/matrix-org/matrix-doc/pull/2867
strongly related, but not exactly the same https://github.com/vector-im/element-web/issues/4406
already released within [SchildiChat](https://github.com/SchildiChat/matrix-react-sdk/pull/2)

## Description
This PR...
- ...adds a new button for marking a room as unread
- ...shows a visual indicator based on `com.famedly.marked_unread`
- ...cleares the `com.famedly.marked_unread` when the room gets opened

![image](https://user-images.githubusercontent.com/10246027/124386947-0eda4200-dcdd-11eb-80b6-b02b709d07aa.png)
![image](https://user-images.githubusercontent.com/10246027/124386967-231e3f00-dcdd-11eb-9a9d-61155ef4bc2c.png)

## Questions
since I'm still quite unfamiliar with the codebase, I got some questions about this PR:
1. ~~should this react to the stable prefix `m.marked_unread` already as well? or is the usual approach to add it as soon as the MSC is in a further state?~~
2. the current code feels like too much is duplicated but can't really say where to rather put the setting & retrieval of the unread state. maybe even put part of in into https://github.com/matrix-org/matrix-js-sdk? would be happy about any hint :)
3. ~~Any suggestions for an fitting icon?~~

<!-- CHANGELOG_PREVIEW_START -->
---
This PR currently has no changelog labels, so will not be included in changelogs.

A reviewer can add one of: `T-Deprecation`, `T-Enhancement`, `T-Defect`, `T-Task` to indicate what type of change this is, or add `Type: [enhancement/defect/task]` to the description and I'll add them for you.<!-- CHANGELOG_PREVIEW_END -->

SimonBrandner reviewed (changes requested):
Thank you for working on this! Generally, looks great! Just some minor details

<hr></hr>

Could you please include a sign-off as described [here](https://github.com/matrix-org/matrix-js-sdk/blob/develop/CONTRIBUTING.rst#sign-off)?

Also, I think this should probably be hidden under a [labs flag](https://github.com/vector-im/element-web/blob/develop/docs/labs.md) since it is using an unstable prefix

<hr></hr>

> should this react to the stable prefix m.marked_unread already as well? or is the usual approach to add it as soon as the MSC is in a further state?

As far as I know, this usually isn't done.

> the current code feels like too much is duplicated but can't really say where to rather put the setting & retrieval of the unread state. maybe even put part of in into matrix-org/matrix-js-sdk? would be happy about any hint :)

I probably wouldn't put it into the `js-sdk` but you could either create a new file for this or put it into an existing one (e.g. `Unread.ts`)

alangecker commented on src/i18n/strings/en_EN.json:
```diff
@@ -1038,6 +1038,7 @@
     "Failed to copy": "Failed to copy",
     "Share invite link": "Share invite link",
     "Invite people": "Invite people",
+    "Mark as unread":"Mark as unread",
```
ah, I was not aware about the script. reverted my change now and executed the script :)

alangecker commented on src/stores/notifications/RoomNotificationState.ts:
```diff
@@ -110,13 +116,21 @@ export class RoomNotificationState extends NotificationState implements IDestroy
+                const isMarkedUnread = this.room.getAccountData('com.famedly.marked_unread')?.getContent()?.unread
+
+                if (isMarkedUnread) {
+                    this._color = NotificationColor.Red;
```
I thought about the bold as well, but the MSC states:
> When this is true a client should show the room with an indeterminate unread marker. This marker should
be of similar visual importance to a notification badge without the notification count

The Notification badge is usually in red, so I assume, that this means the unread count should also be red?

SimonBrandner replied:
Yeah, this makes sense 👍

turt2live replied:
A notification badge is grey or red depending on various states, which translates to bold or red.

alangecker commented:
thanks a lot for the feedback! :)

- I signed off my commits now
- the feature is now behind a feature flag: `feature_mark_unread`
- I hope I resolved all your other points as well :)

SimonBrandner commented on src/stores/notifications/RoomNotificationState.ts:
```diff
@@ -76,6 +87,10 @@ export class RoomNotificationState extends NotificationState implements IDestroy
     };
 
+    private handleRoomAccountDataUpdate = (ev: MatrixEvent) => {
+        this.updateNotificationState();
```
We should probably only do this if this is the relevant event type

SimonBrandner commented on src/stores/notifications/RoomNotificationState.ts:
```diff
@@ -24,16 +24,25 @@ import { Room } from "matrix-js-sdk/src/models/room";
+import SettingsStore from "../../settings/SettingsStore";
 
 export class RoomNotificationState extends NotificationState implements IDestroyable {
+    settingWatcherRef = null;
```
Could we please give this a less generic name and make it private?

SimonBrandner reviewed (commented):
Two more tiny things 😀

alangecker commented:
@SimonBrandner
as many as you find! :)

regarding an icon: can I just add any license compatible one or would a design team potentially become unhappy? :D

SimonBrandner reviewed (approved):
LGTM

Regarding the icon: I think you can use any icon that fits the license. Though it's quite probable the design team will suggest their own in the end

turt2live commented:
@alangecker what's the status of this? It's flagged as WIP and has a bit of a confusing history to it. Is it ready for review by our teams?

alangecker commented:
@turt2live I still didn't manage to find and integrate an icon, that's why it is still flagged as WIP. From the functionality I wouldn't do any more changes and would be happy about a review :)

also if the design team can imagine deciding for an icon, I would be pleased! :)

turt2live commented on src/Rooms.ts:
```diff
@@ -143,3 +143,19 @@ function guessDMRoomTargetId(room: Room, myUserId: string): string {
     return oldestUser.userId;
 }
+
+export const UNSTABLE_MSC2867_MARKED_UNREAD_TYPE = 'com.famedly.marked_unread';
```
we should be using the js-sdk's `UnstableValue` class.

alangecker replied:
done within https://github.com/matrix-org/matrix-react-sdk/pull/6314/commits/6dcd393181a5b605355d306912914fe45eb27050

turt2live commented on src/stores/notifications/RoomNotificationState.ts:
```diff
@@ -24,16 +24,25 @@ import { Room } from "matrix-js-sdk/src/models/room";
 
 export class RoomNotificationState extends NotificationState implements IDestroyable {
+    private featureMarkedUnreadWatcherRef = null;
     constructor(public readonly room: Room) {
```
missing blank line

alangecker replied:
linter is now happy https://github.com/matrix-org/matrix-react-sdk/pull/6314/commits/63f5ccbf158bbf152551c406688184f929cf5abb

turt2live commented on src/stores/notifications/RoomNotificationState.ts:
```diff
@@ -76,6 +87,12 @@ export class RoomNotificationState extends NotificationState implements IDestroy
         }
     };
 
+    private handleRoomAccountDataUpdate = (ev: MatrixEvent) => {
```
this doesn't feel like the right type for account data: there's a ton of missing fields on account data events.

alangecker replied:
With every occurrence in `matrix-react-sdk` there is `.on("Room.accountData")` used along with `MatrixEvent` as the type for its first argument.

What could be the alternative type? :)

turt2live reviewed (commented):
I'm concerned more generally about the MSC itself as this doesn't give a reliable count for the user to rely upon, so they end up trying to discern the difference between manually marked unread, unread, and mentioned. 

I'll copy these comments to the MSC, but overall I don't feel this is a problem that account data should be solving.

alangecker commented:
seems like there is support for going a different path for that feature (https://github.com/matrix-org/matrix-doc/pull/2867#pullrequestreview-723783823), so I won't continue here for now.

I'm still motivated working on an implementation for this feature, but would need others to lead the spec discussions! :)

SpiritCroc commented:
@alangecker thank you for this PR! We have included it in SchildiChat-web/desktop: https://github.com/SchildiChat/matrix-react-sdk/pull/2
Do you want to be mentioned in the changelog?

alangecker commented:
@SpiritCroc I was a bit sad to stop the development, wanted to wait for the spec discussions first, but it's very nice that it's being used now nevertheless! :) 🎉

I would be pleased about credits in changelogs, but I'm also totally fine without! ;)

alangecker commented:
since there is currently no development on the MSC, I thought about pushing this further :)

#### things done
- I've rebased now the branch on top of the current develop and added the additional commits being done by @SpiritCroc  and @su-ex from the [SchildiChat branch](https://github.com/SchildiChat/matrix-react-sdk/pull/2), except keeping this feature still disabled by default.
- resolving most of the open conversations

#### open questions
1. Despite the fact that the MSC is currently still fundamentally in question, can this still be merged behind a feature flag? @turt2live 
2. They commits by @SpiritCroc and @su-ex  are missing a sign-off. Is that an issue and if so how can we resolve that without the complicated path of giving everyone access to this branch and each one has to rebase the commits?^^ couldn't it legally be fine if everyone states here that they sign-off every commit in this PR? :D
3. Two icons from google material design got added https://github.com/matrix-org/matrix-react-sdk/pull/6314/commits/e5810fb3edd9d4072d38ce0f2fa4d25b34a1425e. Is that license wise fine? @SimonBrandner

SpiritCroc commented:
Signed-off-by: Tobias Büttner <dev@spiritcroc.de>

What I know about the license of material design icons I've also written in the commit message of https://github.com/matrix-org/matrix-react-sdk/commit/e5810fb3edd9d4072d38ce0f2fa4d25b34a1425e.

su-ex commented:
Signed-off-by: Quirin Götz <codeworks@supercable.onl>

turt2live commented:
So the story on this is that it's in a bit of a questionable bit spec-wise. I think before we commit to maintenance of this feature we'd want to see a slightly more stable spec discussion first. For the time being, I'm going to close this because it's bitrotting beyond reasonable repair, but it would still count as an implementation against the MSC if needed (though there are plenty of other implementations on the MSC as well). The code would still be here for someone to pick up if needed down the line.

Thanks for taking the time to look at it, and apologies for the long turnaround.
